File: src/parser/lexer.ts
import { Token, TokenType, ParserError } from "./types";

export class Lexer {
    private pos = 0;
    private char = "";
    private line = 1;
    private col = 1;

    private wasmLexer: any;

    constructor(private sql: string, wasmModule?: any) {
        if (wasmModule && wasmModule.Lexer) {
            this.wasmLexer = new wasmModule.Lexer(sql);
        } else {
            this.char = sql[0] || "";
        }
    }

    private advance() {
        if (this.char === '\n') {
            this.line++;
            this.col = 1;
        } else {
            this.col++;
        }
        this.pos++;
        this.char = this.pos < this.sql.length ? this.sql[this.pos]! : "";
    }

    public nexToken(): Token {
        if (this.wasmLexer) {
            const token = this.wasmLexer.nextToken();
            // Emscripten enums might be returned as objects with a .value property or similar.
            // We want the numeric value to match TS TokenType enum.
            const type = typeof token.type === 'object' ? token.type.value : token.type;
            return {
                type: type,
                value: token.value,
                line: token.line,
                column: token.column
            };
        }

        while (this.char && /\s/.test(this.char)) this.advance();
        if (!this.char) return {
            type: TokenType.EOF, value: "", line: this.line, column: this.col
        }

        const startLine = this.line;
        const startCol = this.col;

        if (/\d/.test(this.char)) {
            let value = "";
            while (/\d/.test(this.char)) {
                value += this.char;
                this.advance();
            }
            return {
                type: TokenType.Number, value, line: startLine, column: startCol
            }
        }

        if (/[a-zA-Z_]/.test(this.char)) {
            let value = "";
            while (/[a-zA-Z0-9_]/.test(this.char)) {
                value += this.char;
                this.advance();
            }
            return {
                type: TokenType.Identifier, value, line: startLine, column: startCol
            }
        }

        if (this.char === "'" || this.char === '"') {
            const quote = this.char;
            let value = "";
            this.advance();
            while (this.char && this.char !== quote) {
                value += this.char;
                this.advance();
            }
            if (this.char === quote) {
                this.advance();
            } else {
                throw new ParserError("Unterminated string literal", startLine, startCol);
            }
            return {
                type: TokenType.String, value, line: startLine, column: startCol
            }
        }

        // Check for operators
        const opChars = ['=', '>', '<', '!', '|', '*', '+', '-', '/', '%'];
        if (opChars.includes(this.char)) {
            let val = this.char;
            this.advance();
            if ((val === '=' || val === '>' || val === '<' || val === '!') && this.char === '=') {
                val += this.char;
                this.advance();
            } else if (val === '|' && this.char === '|') {
                val += this.char;
                this.advance();
            } else if (val === '<' && this.char === '>') {
                val += this.char;
                this.advance();
            }
            return {
                type: TokenType.Operator, value: val, line: startLine, column: startCol
            };
        }

        const val = this.char;
        this.advance();
        return {
            type: TokenType.Punctuation, value: val, line: startLine, column: startCol
        }
    }

    public dispose() {
        if (this.wasmLexer && this.wasmLexer.delete) {
            this.wasmLexer.delete();
        }
    }
}

File: src/cpp/lexer.cpp
#include "lexer.hpp"
#include <cctype>
#include <stdexcept>

namespace toy {

Lexer::Lexer(const std::string& sql) : sql(sql), pos(0), line(1), col(1) {
    currentChar = sql.empty() ? '\0' : sql[0];
}

void Lexer::advance() {
    if (currentChar == '\n') {
        line++;
        col = 1;
    } else {
        col++;
    }
    pos++;
    currentChar = (pos < sql.length()) ? sql[pos] : '\0';
}

void Lexer::skipWhitespace() {
    while (currentChar != '\0' && std::isspace(static_cast<unsigned char>(currentChar))) {
        advance();
    }
}

Token Lexer::nextToken() {
    skipWhitespace();

    if (currentChar == '\0') {
        return {TokenType::EOF_TOKEN, "", line, col};
    }

    int startLine = line;
    int startCol = col;

    if (std::isdigit(static_cast<unsigned char>(currentChar))) {
        return handleNumber();
    }

    if (std::isalpha(static_cast<unsigned char>(currentChar)) || currentChar == '_') {
        return handleIdentifier();
    }

    if (currentChar == '\'' || currentChar == '"') {
        return handleString();
    }

    // Handle operators and punctuation
    const std::string opChars = "=><!|*+-/%";
    if (opChars.find(currentChar) != std::string::npos) {
        return handleOperator();
    }

    // Punctuation
    std::string val(1, currentChar);
    advance();
    return {TokenType::Punctuation, val, startLine, startCol};
}

Token Lexer::handleNumber() {
    int startLine = line;
    int startCol = col;
    std::string value;
    while (currentChar != '\0' && std::isdigit(static_cast<unsigned char>(currentChar))) {
        value += currentChar;
        advance();
    }
    return {TokenType::Number, value, startLine, startCol};
}

Token Lexer::handleIdentifier() {
    int startLine = line;
    int startCol = col;
    std::string value;
    while (currentChar != '\0' && (std::isalnum(static_cast<unsigned char>(currentChar)) || currentChar == '_')) {
        value += currentChar;
        advance();
    }
    return {TokenType::Identifier, value, startLine, startCol};
}

Token Lexer::handleString() {
    char quote = currentChar;
    int startLine = line;
    int startCol = col;
    std::string value;
    advance(); // skip quote
    while (currentChar != '\0' && currentChar != quote) {
        value += currentChar;
        advance();
    }

    if (currentChar == quote) {
        advance(); // skip quote
    } else {
        throw std::runtime_error("Unterminated string literal at line " + std::to_string(startLine) + ", col " + std::to_string(startCol));
    }

    return {TokenType::String, value, startLine, startCol};
}

Token Lexer::handleOperator() {
    int startLine = line;
    int startCol = col;
    std::string val(1, currentChar);
    advance();

    if ((val == "=" || val == ">" || val == "<" || val == "!") && currentChar == '=') {
        val += currentChar;
        advance();
    } else if (val == "|" && currentChar == '|') {
        val += currentChar;
        advance();
    } else if (val == "<" && currentChar == '>') {
        val += currentChar;
        advance();
    }

    return {TokenType::Operator, val, startLine, startCol};
}

} // namespace toy
